### ISSUE TITLE
Unexpected behavior with @include / @skip directives

### ISSUE THREAD

**Issue by vladar** [facebook/relay#507]
I have following fragment: 

```
fragment on Story {
  id
  comments @include(if: $showComments) {
    id
    text
  }
}
```

Then somewhere in the app I have a list of `Stories`. When `$showComments` is `true` everything works as expected. But when `$showComments` is `false` Relay sends additional GraphQL query for each story after initial load.

This query is sent to `Node` interface and includes this Story fragment (+child fragments if any) as well as `id` and `__typename` fields.

The response returned is always the same:

```
{
  "node": {
    "__typename": "Story",
    "id": "node-id-value"
  }
}
```

Same applies to `@skip(if: true)` directive.

**Comment by josephsavona** [comment]
Relay currently passes through directives to the server but does not directly implement support for `@include` or `@skip`. We should add support such that:

- unincluded fields (include: false or skip: true) are removed during diffing
- reader does not return unincluded fields

Contributions welcome, otherwise we'll get to this as time permits.

**Comment by josephsavona** [comment]
merged into #536
